### ISSUE TITLE
Support for relationships

### ISSUE THREAD

**Issue by axelgenus** [rethinkdb/rethinkdb-go#275]
Right now Gorethink supports nested structs and stores them as document/sub-document. In some complex cases though it is better to use plain old relationships using ID's (one-to-many). RethinkDB does support JOIN queries but it is impossible with Gorethink to map nested structs with document references.

I prepared an [example](https://gist.github.com/axelgenus/8170f009175cdd00e419). Run it and check the result on the RethinkDB Administration Console's Data Explorer:

r.db("test").table("writers")

``` json
{
    "id":  "c2182a10-6b9d-4ea1-a70c-d6649bb5f8d7" ,
    "name":  "JRR Tolkien"
}
```

r.db("test").table("books")

``` json
{
    "author": {
        "name":  "JRR Tolkien"
    } ,
    "id":  "eeb006d6-7fec-46c8-9d29-45b83f07ca14",
    "title":  "The Hobbit"
}
{
    "author": {
        "name":  "JRR Tolkien"
    } ,
    "id":  "ca882bde-b4c7-42d2-b0c3-11e07ccd15e3",
    "title":  "The Lord of the Rings"
}
```

This is weird: Gorethink copied the "author" data as a sub-document in the "book" documents. Later on, if I want to change the name of the author (from "JRR Tolkien" to "J.R.R. Tolkien") I have to retrive the author, modify the object, retrive all the books and reassign the "author" reference with the modified object.

The relational approach can be obtained implementing Marshaler/Unmarshaler interfaces in the objects that needs to store ID references instead of the copy of the referenced object's data. My proposal is to enhance Gorethink adding a keyword in the annotations of the struct's field like this:

``` go
type Writer struct {
    ID      string  `gorethink:"id,omitempty"`
    Name    string  `gorethink:"name"`
}

type Book struct {
    ID      string  `gorethink:"id,omitempty"`
    Title   string  `gorethink:"title"`
    Author  *Writer `gorethink:"author_id,reference"` // <-----
}
```

In case the default marshaller finds this kind of annotation then it can use the writer's ID instead of copying the writer's data in the books' documents. The result should be something like this:

r.db("test").table("books")

``` json
{
    "author_id": "c2182a10-6b9d-4ea1-a70c-d6649bb5f8d7",
    "id":  "eeb006d6-7fec-46c8-9d29-45b83f07ca14",
    "title":  "The Hobbit"
}
{
    "author_id": "c2182a10-6b9d-4ea1-a70c-d6649bb5f8d7",
    "id":  "ca882bde-b4c7-42d2-b0c3-11e07ccd15e3",
    "title":  "The Lord of the Rings"
}
```

Also, when retrieving the documents from the database, Gorethink already knows that the "Author" field is a reference (because we tagged the field) and it can either perform a merge query to immediately get also the referenced document or get it later in a "lazy" fashion.

**Comment by dancannon** [comment]
Hey, thanks for the suggestion. I will look into it adding this feature today (not sure how far I will get though).

**Comment by dancannon** [comment]
I have pushed some code to the develop branch which should hopefully do what you want however I decided to not change the retrieval of the referenced document since I dont really want to start changing user queries. 

https://github.com/dancannon/gorethink/tree/develop#references

**Comment by axelgenus** [comment]
I'll dig into this ASAP and I'll let you know. You've been astonishingly fast! O.O

**Comment by axelgenus** [comment]
Nice! It seems to work just fine. The encoder part seems to be ok but what about the decoder? If it finds the "reference" tag, does it just ignore it?

**Comment by dancannon** [comment]
Currently I have left it up to the user to create the selection query. Assuming the query is correct (I have added an example of such a query to the README) the library should decode the results without any further changes.

**Comment by dancannon** [comment]
This has now been released as part of 1.4.0
